File: tools/opai-dev/services/supabase.js
/**
 * OPAI Dev — Supabase REST API wrapper for dev_workspaces table.
 * Uses service key (bypasses RLS) via node-fetch.
 */

const fetch = require('node-fetch');

const SUPABASE_URL = process.env.SUPABASE_URL || '';
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY || '';
const TABLE = 'dev_workspaces';

function headers() {
  return {
    apikey: SUPABASE_SERVICE_KEY,
    Authorization: `Bearer ${SUPABASE_SERVICE_KEY}`,
    'Content-Type': 'application/json',
    Prefer: 'return=representation',
  };
}

function url(query = '') {
  return `${SUPABASE_URL}/rest/v1/${TABLE}${query}`;
}

async function createWorkspace(data) {
  const resp = await fetch(url(), {
    method: 'POST',
    headers: headers(),
    body: JSON.stringify(data),
  });
  if (!resp.ok) throw new Error(`Supabase insert failed: ${resp.status} ${await resp.text()}`);
  const rows = await resp.json();
  return rows[0];
}

async function getWorkspace(id) {
  const resp = await fetch(url(`?id=eq.${id}&select=*`), { headers: headers() });
  if (!resp.ok) throw new Error(`Supabase select failed: ${resp.status}`);
  const rows = await resp.json();
  return rows[0] || null;
}

async function getWorkspacesByUser(userId) {
  const resp = await fetch(
    url(`?user_id=eq.${userId}&status=not.eq.destroyed&order=created_at.desc`),
    { headers: headers() }
  );
  if (!resp.ok) throw new Error(`Supabase select failed: ${resp.status}`);
  return resp.json();
}

async function getActiveWorkspaces() {
  const resp = await fetch(
    url(`?status=in.(creating,running,stopped)&select=*`),
    { headers: headers() }
  );
  if (!resp.ok) throw new Error(`Supabase select failed: ${resp.status}`);
  return resp.json();
}

async function updateWorkspace(id, data) {
  const resp = await fetch(url(`?id=eq.${id}`), {
    method: 'PATCH',
    headers: headers(),
    body: JSON.stringify(data),
  });
  if (!resp.ok) throw new Error(`Supabase update failed: ${resp.status} ${await resp.text()}`);
  const rows = await resp.json();
  return rows[0];
}

async function updateLastActive(id) {
  return updateWorkspace(id, { last_active_at: new Date().toISOString() });
}

async function getRunningWorkspaces() {
  const resp = await fetch(
    url(`?status=eq.running&select=*`),
    { headers: headers() }
  );
  if (!resp.ok) throw new Error(`Supabase select failed: ${resp.status}`);
  return resp.json();
}

async function getStoppedWorkspaces() {
  const resp = await fetch(
    url(`?status=eq.stopped&select=*`),
    { headers: headers() }
  );
  if (!resp.ok) throw new Error(`Supabase select failed: ${resp.status}`);
  return resp.json();
}

module.exports = {
  createWorkspace,
  getWorkspace,
  getWorkspacesByUser,
  getActiveWorkspaces,
  updateWorkspace,
  updateLastActive,
  getRunningWorkspaces,
  getStoppedWorkspaces,
};

File: tools/opai-studio/core/supabase.py
"""OPAI Studio — Supabase REST API helpers (async httpx)."""

import logging
import httpx
import config

log = logging.getLogger("studio.supabase")


def _headers(token=None):
    headers = {
        "apikey": config.SUPABASE_SERVICE_KEY,
        "Authorization": f"Bearer {config.SUPABASE_SERVICE_KEY}",
        "Content-Type": "application/json",
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


async def sb_get(path, token=None):
    """GET from Supabase REST. path includes table + query string."""
    async with httpx.AsyncClient(timeout=15) as c:
        r = await c.get(
            f"{config.SUPABASE_URL}/rest/v1/{path}",
            headers=_headers(token),
        )
        r.raise_for_status()
        return r.json()


async def sb_post(path, data, token=None):
    """POST to Supabase REST. Returns created row."""
    headers = _headers(token)
    headers["Prefer"] = "return=representation"
    async with httpx.AsyncClient(timeout=15) as c:
        r = await c.post(
            f"{config.SUPABASE_URL}/rest/v1/{path}",
            headers=headers,
            json=data,
        )
        if not r.is_success:
            log.error("POST %s → %d: %s", path, r.status_code, r.text[:300])
        r.raise_for_status()
        rows = r.json()
        return rows[0] if isinstance(rows, list) and rows else rows


async def sb_patch(path, data, token=None):
    """PATCH Supabase REST. path should include filters."""
    headers = _headers(token)
    headers["Prefer"] = "return=representation"
    async with httpx.AsyncClient(timeout=30) as c:
        r = await c.patch(
            f"{config.SUPABASE_URL}/rest/v1/{path}",
            headers=headers,
            json=data,
        )
        r.raise_for_status()
        rows = r.json()
        return rows[0] if isinstance(rows, list) and rows else rows


async def sb_delete(path, token=None):
    """DELETE from Supabase REST. path should include filters."""
    headers = _headers(token)
    headers["Prefer"] = "return=minimal"
    async with httpx.AsyncClient(timeout=15) as c:
        r = await c.delete(
            f"{config.SUPABASE_URL}/rest/v1/{path}",
            headers=headers,
        )
        r.raise_for_status()
